modules/layouts/CompressedSparseLayout.chpl

```diff
@@ -0,0 +1,979 @@
+                hi = stopIdx(c);
+          for r in lo..hi {
+            f.write(" (", idx(r), ", ", c, ")");
+            if r == hi then f.write("\n");
+          }
+        }
+      }
+      f.write("}\n");
+    }
+
+    override proc dsiSupportsAutoLocalAccess() param {
+      return csLayoutSupportsAutoLocalAccess;
+    }
+
+  } // CSDom
+
+
+  @chpldoc.nodoc
+  proc CSDom.rows() {
+    return this.rowRange;
+  }
+
+  @chpldoc.nodoc
+  proc CSDom.cols() {
+    return this.colRange;
+  }
+
+  @chpldoc.nodoc
+  iter CSDom.uidsInRowCol(rc) {
+    for uid in startIdx[rc]..<startIdx[rc+1] do
+      yield uid;
+  }
+
+
+  @chpldoc.nodoc
+  class CSArr: BaseSparseArrImpl(?) {
+
+    proc init(type eltType,
+              param rank : int,
+              type idxType,
+              dom,
+              param initElts:bool) {
+      super.init(eltType, rank, idxType, dom, initElts);
+    }
+
+    // dsiDestroyArr is defined in BaseSparseArrImpl
+
+    proc dsiAccess(ind: rank*idxType) ref {
+      // make sure we're in the dense bounding box
+      dom.boundsCheck(ind);
+
+      // lookup the index and return the data or IRV
+      const (found, loc) = dom.find(ind);
+      if found then
+        return data(loc);
+      else
+        halt("attempting to assign a 'zero' value in a sparse array at index ",
+             if rank == 1 then ind(0) else ind);
+    }
+    // value version for POD types
+    proc dsiAccess(ind: rank*idxType)
+      where shouldReturnRvalueByValue(eltType) {
+      // make sure we're in the dense bounding box
+      dom.boundsCheck(ind);
+
+      // lookup the index and return the data or IRV
+      const (found, loc) = dom.find(ind);
+      if found then
+        return data(loc);
+      else
+        return irv;
+    }
+    // const ref version for types with copy ctors
+    proc dsiAccess(ind: rank*idxType) const ref {
+      // make sure we're in the dense bounding box
+      dom.boundsCheck(ind);
+
+      // lookup the index and return the data or IRV
+      const (found, loc) = dom.find(ind);
+      if found then
+        return data(loc);
+      else
+        return irv;
+    }
+
+
+
+    iter these() ref {
+      foreach i in 0..#dom._nnz do yield data[i];
+    }
+
+    iter these(param tag: iterKind) where tag == iterKind.leader {
+      // forward to the leader iterator on our domain
+      // Note: this is so that arrays can be zippered with domains;
+      // otherwise just chunk up data[1..dom._nnz] a-la DefaultSparseArr.
+      for followThis in dom.these(tag) do
+        yield followThis;
+    }
+
+    iter these(param tag: iterKind, followThis: (?,?,?)) ref where tag == iterKind.follower {
+      // simpler than CSDom's follower - no need to deal with rows (or columns)
+      var (followThisDom, startIx, endIx) = followThis;
+
+      if (followThisDom != this.dom) then
+        halt("Sparse arrays can't be zippered with anything other than their domains and sibling arrays (CS layout)");
```

Comment:
Why does this message say (CS layout)? Is it to tell us where the error is coming from? Can we remove it?

Reply:
I think that's a holdover from the previous module that I didn't notice or think to update, thanks.  I'll check for other uses of 'CS' that may have slipped through as well (I thought I'd already done that, but…).